File: src/db-toolkit/core/routes/analytics.py
"""Analytics routes for database monitoring."""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from operations.analytics_manager import AnalyticsManager
from operations.connection_manager import ConnectionManager

router = APIRouter()
connection_manager = ConnectionManager()


class KillQueryRequest(BaseModel):
    """Kill query request."""
    pid: int


class QueryPlanRequest(BaseModel):
    """Query plan request."""
    query: str


@router.get("/connections/{connection_id}")
async def get_analytics(connection_id: int):
    """Get database analytics."""
    try:
        connection_info = connection_manager.get_connection(connection_id)
        if not connection_info:
            raise HTTPException(status_code=404, detail="Connection not found")
        
        config, connector = connection_info
        analytics_manager = AnalyticsManager(connector.connection)
        result = await analytics_manager.get_analytics(config, connection_id)
        
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/connections/{connection_id}/history")
async def get_historical_data(connection_id: int, hours: int = Query(default=3, ge=1, le=3)):
    """Get historical analytics data."""
    try:
        connection_info = connection_manager.get_connection(connection_id)
        if not connection_info:
            raise HTTPException(status_code=404, detail="Connection not found")
        
        config, connector = connection_info
        analytics_manager = AnalyticsManager(connector.connection)
        history = analytics_manager.get_historical_data(connection_id, hours)
        
        return {"success": True, "history": history}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/connections/{connection_id}/query-plan")
async def get_query_plan(connection_id: int, request: QueryPlanRequest):
    """Get query execution plan."""
    try:
        connection_info = connection_manager.get_connection(connection_id)
        if not connection_info:
            raise HTTPException(status_code=404, detail="Connection not found")
        
        config, connector = connection_info
        analytics_manager = AnalyticsManager(connector.connection)
        result = await analytics_manager.get_query_plan(request.query, config)
        
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/connections/{connection_id}/kill")
async def kill_query(connection_id: int, request: KillQueryRequest):
    """Kill a running query."""
    try:
        connection_info = connection_manager.get_connection(connection_id)
        if not connection_info:
            raise HTTPException(status_code=404, detail="Connection not found")
        
        config, connector = connection_info
        analytics_manager = AnalyticsManager(connector.connection)
        result = await analytics_manager.kill_query(request.pid, config)
        
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: src/db-toolkit/electron/backend/handlers/analytics.js
/**
 * IPC handlers for analytics operations.
 */

const { ipcMain, shell, app } = require('electron');
const path = require('path');
const fs = require('fs').promises;
const { connectionManager } = require('../utils/connection-manager');
const { connectionStorage } = require('../utils/connection-storage');
const { AnalyticsManager } = require('../operations/analytics-manager');

function registerAnalyticsHandlers() {
  ipcMain.handle('analytics:get', async (event, connectionId) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const config = await connectionStorage.getConnection(connectionId);
      const manager = new AnalyticsManager(connection);
      return await manager.getAnalytics(config, connectionId);
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:historical', async (event, connectionId, hours = 3) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const manager = new AnalyticsManager(connection);
      return { success: true, data: manager.getHistoricalData(connectionId, hours) };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:slow-queries', async (event, connectionId, hours = 24) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const manager = new AnalyticsManager(connection);
      return { success: true, data: manager.getSlowQueryLog(connectionId, hours) };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:table-stats', async (event, connectionId) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const config = await connectionStorage.getConnection(connectionId);
      const manager = new AnalyticsManager(connection);
      return { success: true, data: await manager.getTableStatistics(config) };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:export-pdf', async (event, connectionId) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const config = await connectionStorage.getConnection(connectionId);
      const manager = new AnalyticsManager(connection);
      const pdfBuffer = await manager.exportToPDF(connectionId, config.name, config);
      
      const downloadsPath = app.getPath('downloads');
      const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
      const filename = `analytics-${config.name}-${timestamp}.pdf`;
      const filepath = path.join(downloadsPath, filename);
      
      await fs.writeFile(filepath, pdfBuffer);
      shell.showItemInFolder(filepath);
      
      return { success: true, path: filepath };
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:query-plan', async (event, connectionId, query) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const config = await connectionStorage.getConnection(connectionId);
      const manager = new AnalyticsManager(connection);
      return await manager.getQueryPlan(query, config);
    } catch (error) {
      return { success: false, error: error.message };
    }
  });

  ipcMain.handle('analytics:kill-query', async (event, connectionId, pid) => {
    try {
      const connection = await connectionManager.getConnector(connectionId);
      if (!connection) {
        return { success: false, error: 'Connection not found' };
      }

      const config = await connectionStorage.getConnection(connectionId);
      const manager = new AnalyticsManager(connection);
      return await manager.killQuery(pid, config);
    } catch (error) {
      return { success: false, error: error.message };
    }
  });
}

module.exports = { registerAnalyticsHandlers };
